core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -156,6 +160,16 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                 }
             );
         }
+
+        private void handleShardFailureOnMaster(final ShardRoutingEntry shardRoutingEntry, ClusterStateTaskListener listener) {
```

Comment:
> does this mean the class can be static (if we give it a logger) ?

@bleskes No, because we also capture instances of `ShardFailedClusterStateHandler` and `ShardStartedClusterStateHandler`. Inlining those into fields of the respective transport handlers would mean that they need to be static as well, but they capture fields from `ShardStateAction`. I could go either way on whether or not it's worth the trouble to chain everything through but I pushed 7591f2047a1f46f0ccce44ec0a1e1cbdd6598b2d so that you can help decide?